### PR TITLE
Fix a false positive for `FactoryBot/AssociationStyle` when `EnforcedStyle: Explicit` and using trait within trait

### PR DESCRIPTION
**Pull request by ydah**
Fix: https://github.com/rubocop/rubocop-factory_bot/issues/54

______________________________________________________________________

Before submitting the PR make sure the following are checked:

- [x] Feature branch is up-to-date with `master` (if not - rebase it).
- [x] Squashed related commits together.
- [x] Added tests.
- [-] Updated documentation.
- [x] Added an entry to the `CHANGELOG.md` if the new code introduces user-observable changes.
- [x] The build (`bundle exec rake`) passes (be sure to run this locally, since it may produce updated documentation that you will need to commit).

### CODE REVIEW

File: lib/rubocop/cop/factory_bot/association_style.rb [hunk 1]
```diff
@@ -218,6 +224,14 @@ def options_for_autocorrect_to_implicit_style(node)
           end
           options
         end
+
+        def trait_within_trait?(node)
+          node.each_ancestor(:block).any? do |ancestor|
+            if ancestor.send_node.method?(:factory)
+              trait_name(ancestor).include?(node.method_name)
+            end
+          end
```
Comment (Darhazer) [inline lib/rubocop/cop/factory_bot/association_style.rb:233]: Performance wise, shouldn't we compile a list of traits on the very definition of the factory, and then reuse it. Or just go to the top block, as now you're traveling part of the tree every time you are going up (when you want to look also into the siblings)

Comment (ydah) [reply lib/rubocop/cop/factory_bot/association_style.rb:233]: I updated this PR. Thank you.

File: lib/rubocop/cop/factory_bot/association_style.rb [hunk 2]
```diff
@@ -158,7 +163,8 @@ def autocorrect_to_implicit_style(corrector, node)
 
         def bad?(node)
           if style == :explicit
-            implicit_association?(node)
+            implicit_association?(node) &&
```
Comment (pirj) [inline lib/rubocop/cop/factory_bot/association_style.rb:166]: Cosmetic. Here it feels that ‘implicit_association?’ can return true when it only looks like an implicit association, not necessarily is.

Comment (ydah) [reply lib/rubocop/cop/factory_bot/association_style.rb:166]: Maybe we should rename the matcher? I can't think of a good name now 🤔

File: lib/rubocop/cop/factory_bot/association_style.rb [hunk 3]
```diff
@@ -218,6 +224,14 @@ def options_for_autocorrect_to_implicit_style(node)
           end
           options
         end
+
+        def trait_within_trait?(node)
+          factory_node = node.each_ancestor(:block).find do |ancestor|
```
Comment (pirj) [inline lib/rubocop/cop/factory_bot/association_style.rb:229]: What if it’s a factory inside a factory and the trait used in the inner one is defined in the outer one?

File: lib/rubocop/cop/factory_bot/association_style.rb [hunk 4]
```diff
@@ -218,6 +224,14 @@ def options_for_autocorrect_to_implicit_style(node)
...
+            ancestor.send_node.method?(:factory)
```
Comment (pirj) [inline lib/rubocop/cop/factory_bot/association_style.rb:230]: I’m worried about the time complexity of this.

File: lib/rubocop/cop/factory_bot/association_style.rb (hunk 3 above)
Comment (ydah) [reply lib/rubocop/cop/factory_bot/association_style.rb]: Indeed it is, I updated this PR to take into account the case of factory inside a factory.

File: lib/rubocop/cop/factory_bot/association_style.rb (hunk 4 above)
Comment (ydah) [reply lib/rubocop/cop/factory_bot/association_style.rb]: The logic is modified for https://github.com/rubocop/rubocop-factory_bot/pull/59#discussion_r1292613483 . Please point out if there is another way to do it better.

**Comment by ydah** [comment]
I updated this PR.

**Review by pirj — APPROVED** [review]
Looks good.
Have you had a chance to run it against a real codebase?

**Comment by ydah** [comment]
I ran `bundle exec rubocop --only FactoryBot/AssociationStyle` on a large Rails project I have on hand and did not encounter any errors or false positives.

```
11593 files inspected, no offenses detected
```

**Pull request merged**